sqlglot/diff.py

```diff
@@ -0,0 +1,300 @@
+                    ) or leaf_similarity_score >= 0.8:
+                        matching_set.add((source_node_id, target_node_id))
+                        self._unmatched_source_nodes.remove(source_node_id)
+                        self._unmatched_target_nodes.remove(target_node_id)
+                        ordered_unmatched_target_nodes.pop(target_node_id, None)
+                        break
+
+        return matching_set
+
+    def _compute_leaf_matching_set(self):
+        leaf_matchings = []
+        for source_leaf in _get_leaves(self._source):
+            for target_leaf in _get_leaves(self._target):
+                if _is_same_label(source_leaf, target_leaf):
+                    similarity_score = _dice_coefficient(
+                        source_leaf.sql(), target_leaf.sql()
+                    )
+                    if similarity_score and similarity_score >= self.f:
+                        leaf_matchings.append(
+                            (source_leaf, target_leaf, similarity_score)
+                        )
+
+        # Pick best matchings based on the highest score
+        matching_set = set()
+        leaf_matchings = sorted(leaf_matchings, key=lambda x: -x[2])
+        for source_leaf, target_leaf, _ in leaf_matchings:
+            if (
+                id(source_leaf) in self._unmatched_source_nodes
+                and id(target_leaf) in self._unmatched_target_nodes
+            ):
+                matching_set.add((id(source_leaf), id(target_leaf)))
+                self._unmatched_source_nodes.remove(id(source_leaf))
+                self._unmatched_target_nodes.remove(id(target_leaf))
+
+        return matching_set
+
+
+def _get_leaves(expression):
+    return expression.find_all(*LEAF_EXPRESSION_TYPES)
+
+
+def _is_same_label(source, target):
+    if _all_same_type(Join, source, target):
+        return source.args.get("side") == target.args.get("side")
+
+    return type(source) is type(target)
+
+
+def _dice_coefficient(string_a, string_b, n=2):
+    n_grams_a = _n_grams(string_a.lower(), n)
+    n_grams_b = _n_grams(string_b.lower(), n)
+
+    total_grams = len(n_grams_a) + len(n_grams_b)
+    if total_grams == 0:
+        return 1.0 if string_a == string_b else 0.0
+
+    overlap_len = 0
+    grams_histo = defaultdict(int)
+    for g in n_grams_a:
+        grams_histo[g] += 1
+    for g in n_grams_b:
+        if grams_histo[g] > 0:
+            overlap_len += 1
+        grams_histo[g] -= 1
+
+    return 2 * overlap_len / total_grams
+
+
+def _n_grams(sequence, n):
+    count = max(0, len(sequence) - n + 1)
+    return [sequence[i : i + n] for i in range(count)]
+
+
+def _expression_only_args(expression):
+    args = []
+    if expression:
+        for a in expression.args.values():
+            args.extend(ensure_list(a))
+    return [a for a in args if isinstance(a, Expression)]
+
+
+def _all_same_type(tpe, *args):
+    return all(isinstance(a, tpe) for a in args)
+
+
+def _lcs(seq_a, seq_b, equal):
```

Comment:
curious why equal is a lambda, seems like ti may not need to be? you only call _lcs in one place right?

Reply:
So the LCS (Least Common Subsequence) is a pretty well-known and generic algorithm and I didn't want to inject our business logic directly into it. I felt like having an abstraction over what's considered to be `equal` is more readable and a better segregation of concerns. 

The alternative was passing the reference to the `matching_set` directly here - https://github.com/tobymao/sqlglot/blob/a4e1a54525ac05ade3dbb73253646cca92ba3c76/sqlglot/diff.py#L137.

Let me know what you think.

Reply:
ok that's fine